### ISSUE TITLE
Smooth handling of deleted or masked TOAs

### ISSUE THREAD

**Issue by paulray** [nanograv/PINT#282]
PINT needs a nice way to handle deleting or masking out TOAs.

If you have a TOAs object, you may want to remove or mask all TOAs with an error bigger than some EMAX value.  Or, in a PINT GUI, you might want to right click on TOAs to delete them.  PINT should then be able to recompute the residuals, or refit the model, with those TOAs deleted from the TOAs table (or, likely better, flagged to be not included in calculations like that).

As far as I can tell, this functionality isn't in PINT.

**Comment by luojing1211** [comment]
@paulray, you are right. It is not there yet. A quick question for deleting TOAs. If a TOA gets deleted, will it still store in the TOA table but not calculation will be applied on it? Or is it just get deleted from TOA tobale period?

**Comment by paulray** [comment]
It could be implemented either way. Both have pros and cons.  I'm not sure which is best.

**Comment by scottransom** [comment]
Note that one of the reasons why we went with astropy tables is that masking is already built in:  http://docs.astropy.org/en/stable/table/masking.html

**Comment by paulray** [comment]
Yeah, masking in Tables should be helpful, but we still need to add some API in the `TOAs()` class.

**Comment by scottransom** [comment]
Not really.  A very simple way of doing that is to generate a mask based on however you want to make it, and then just select a new TOA table that includes all the TOAs that you want to use.  You then do fitting etc with that.  That is basically doing what you want for free by using Python's selection and array functions.

**Comment by paulray** [comment]
@scottransom, how about adding a few lines to `examples/fit_NGC6440E.py` that mask TOAs with errors greater than 50 µs before doing the fit?  Can you do that? It would be great to have a simple example to point to.

**Comment by paulray** [comment]
@scottransom if you do the above, make sure to pull first since I just merged some changes to that script to make it work with ipython and python3

**Comment by scottransom** [comment]
So I just added 3 examples of selecting subsets of TOAs in fit_NGC6440E.py.  The exmaples are all commented out, so you can try them as you like and no tests fail.

There is only one slightly awkward thing.  After selecting, we need to update the .ntoas attribute of the toas class instance.  Maybe if we made that a property that always grabbed the length of the toas.table that issue would go away.

**Comment by paulray** [comment]
Thanks for those. They all use array slicing to grab a subset of the Table, rather than masked arrays.  Do you think that is preferable?  I'm imagining a large NANOGrav data set where someone is looking at the TOAs in a GUI and right clicking to delete one at a time. Should it have to remake the TOAs table with each click?

And we definitely should not have users required to know implementation details like needing to `group_by('obs')`, and fix `ntoas`.  There may be other pieces of data in the `TOAs()` implementation that need to be updated as well, when TOAs are deleted, such as the set of `observatories`. Also you are working only on the table, not the `.toas` list. Could that cause an inconsistency?

So, maybe we add a method to the `TOAs()` class that lets you pass a mask and it deletes the TOAs and does the internal housekeeping?  It might even have an option (`mask=True|False`) that determines whether is uses a masked table or really deletes the TOAs.

**Comment by scottransom** [comment]
Few points...  You don't need to re-make the table with each click.  Only before you do something with the  TOAs.  So if you are using a GUI, it could build up a list of TOAs that you want to remove, and when you do something, the gui re-makes the TOA table once.

The ntoas thing (and the .observatories thing) can be fixed (I'm pretty sure) by a trivial change to make them properties.
And the group_by thing (if we think it is necessary) could be checked and re-grouped if necessary by the model.phase() method.

As for the toas list, that doesn't exist anymore.  That is zapped as soon as a table is created.

I think a simple method to do these things is fine.  But I think it is cool that it basically works now without a method.

Note about masks:  they are basically doing array selection using slices under the covers anyways.  I don't think we need to worry about masking vs returning a new table since we are not talking about much memory usage.  If we did make a selection method, we could have it put the original (or previous) selection in a .table_bak (or similar) table.  That would allow you to instantly undo a selection via copying that to .table.

Bottom line though, is that we want to make it easy to make selections, I think.  And not force people to use a specific selection method.

**Comment by scottransom** [comment]
I'll poke around with these ideas over the next hour or so in a new branch.

**Comment by scottransom** [comment]
OK.  Check out #288

**Comment by scottransom** [comment]
Closing for now.
